### PR TITLE
unet's `sample_size` attribute is to accept tuple(h, w) in `StableDiffusionPipeline`

### LINKED ISSUE #10139: Unet sample_size not fully considered in pipeline init
#### Background
- I'm trying to train a small unet from scratch
- For this, I used a modified version of `train_text_to_image.py`
- unet model's `sample_size` is not symmetric. I passed `[60, 80]` when instantiating `UNet2DConditionModel`, which is normally an `int` value with `512`

#### What happens
- An error occurs in the validation step, using `log_validation()`
- It's because the pipeline(here, `StableDiffusionPipeline`) loaded in the method doesn't expect `unet.config.sample_size` to be a `list`, but an `int` when checking for `deprecation_message`.
- But I think it should be able to accept `list` which is already one of the properly accepted ways to pass a value for `UNet2DConditionModel`'s [`sample_size` parameter](https://huggingface.co/docs/diffusers/en/api/models/unet2d-cond#diffusers.UNet2DConditionModel.sample_size).
- The exact line where the error occurs is the code below
```python
# in `pipeline_stable_diffusion.py` line 258
is_unet_sample_size_less_64 = hasattr(unet.config, "sample_size") and unet.config.sample_size < 64
```

#### Other things to mention
- It seems like that line of code is spread across many other files, such as other `pipeline_stable...py`.
- If you want, I can spend some time on this and make a PR
- If so, I'm thinking of making that line of code with other relevant lines into a function and put that function somewhere such as in `utils`(the location needs to be specified by other members of HF staff)

### Reproduction

```python
from diffusers import StableDiffusionPipeline, UNet2DConditionModel

# load pre-trained model
sd_repo = "stable-diffusion-v1-5/stable-diffusion-v1-5"
pipe = StableDiffusionPipeline.from_pretrained(sd_repo, use_safetensors=True)

# prepare components
small_unet = UNet2DConditionModel(sample_size=[60, 80], block_out_channels=[32, 64, 128, 256])
vae = pipe.components["vae"]
text_encoder = pipe.components["text_encoder"]
tokenizer = pipe.components["tokenizer"]
scheduler = pipe.components["scheduler"]
safety_checker = None
feature_extractor = pipe.components["feature_extractor"]
image_encoder = pipe.components["image_encoder"]

# error occurs(this would be the same when loading with .from_pretrained())
new_pipe = StableDiffusionPipeline(
    vae=vae,
    text_encoder=text_encoder,
    tokenizer=tokenizer,
    unet=small_unet,
    scheduler=scheduler,
    safety_checker=safety_checker,
    feature_extractor=feature_extractor,
    image_encoder=image_encoder
)

```

### Logs

```shell
[rank0]: Traceback (most recent call last):
[rank0]:   File "/root/workspace/03_gen/generative_02/diffusers/examples/text_to_image/train_text_to_image_with_new_unet_from_scratch.py", line 1208, in <module>
[rank0]:     main()
[rank0]:   File "/root/workspace/03_gen/generative_02/diffusers/examples/text_to_image/train_text_to_image_with_new_unet_from_scratch.py", line 1143, in main
[rank0]:     log_validation(
[rank0]:   File "/root/workspace/03_gen/generative_02/diffusers/examples/text_to_image/train_text_to_image_with_new_unet_from_scratch.py", line 144, in log_validation
[rank0]:     pipeline = StableDiffusionPipeline.from_pretrained(
[rank0]:   File "/root/workspace/03_gen/generative_02/.venv/lib/python3.10/site-packages/huggingface_hub/utils/_validators.py", line 114, in _inner_fn
[rank0]:     return fn(*args, **kwargs)
[rank0]:   File "/root/workspace/03_gen/generative_02/.venv/lib/python3.10/site-packages/diffusers/pipelines/pipeline_utils.py", line 948, in from_pretrained
[rank0]:     model = pipeline_class(**init_kwargs)
[rank0]:   File "/root/workspace/03_gen/generative_02/.venv/lib/python3.10/site-packages/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py", line 258, in __init__
[rank0]:     is_unet_sample_size_less_64 = hasattr(unet.config, "sample_size") and unet.config.sample_size < 64
[rank0]: TypeError: '<' not supported between instances of 'list' and 'int'
```
```


### System Info

- 🤗 Diffusers version: 0.32.0.dev0
- Platform: Linux-5.4.0-167-generic-x86_64-with-glibc2.31
- Running on Google Colab?: No
- Python version: 3.10.10
- PyTorch version (GPU?): 2.5.1+cu124 (True)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Huggingface_hub version: 0.26.2
- Transformers version: 4.46.3
- Accelerate version: 1.1.1
- PEFT version: 0.7.0
- Bitsandbytes version: not installed
- Safetensors version: 0.4.5
- xFormers version: not installed
- Accelerator: GRID A100X-10C, 10240 MiB
GRID A100X-10C, 10240 MiB
- Using GPU in script?: <fill in>
- Using distributed or parallel set-up in script?: <fill in>

### Who can help?

@yiyixuxu

### PR DESCRIPTION
**Pull request by Foundsheep**
# What does this PR do?

Fixes #10139
- `UNet2DCondition` model can be passed for `sample_size` parameter with `int` or `Tuple[int, int]`  types in `unet_2d_condition.py`. However, `StableDiffusionPipeline` was not compatible with that aspect and assumed the argument would be only `int` type.
- Now, `StableDiffusionPipeline` can expect `sample_size` for unet which is then to be passed as `height` and `width` to be not just a single number, but also two different numbers in a tuple or list.
- Also modified the difference between docs and type hint to match each other.
- Also added a test case to check if the `StableDiffusionPipeline` would be instantiated without an error in that case.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/diffusers/blob/main/CONTRIBUTING.md)?
- [x] Did you read our [philosophy doc](https://github.com/huggingface/diffusers/blob/main/PHILOSOPHY.md) (important for complex PRs)?
- [x] Was this discussed/approved via a GitHub issue or the [forum](https://discuss.huggingface.co/c/discussion-related-to-httpsgithubcomhuggingfacediffusers/63)? Please add a link to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/diffusers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/diffusers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.
@yiyixuxu @hlky

### CODE REVIEW

**Comment by SahilCarterr** [comment]
Hey @Foundsheep it would be great if you write a test for testing unet's `sample_size`

**Comment by Foundsheep** [comment]
@SahilCarterr thanks. That makes perfect sense, and actually I made my own script to test it and removed it after the test, since the script did not seem to fit with the original code.

I'm wondering, where would the test code be placed in the repository if I should make one?

**Comment by SahilCarterr** [comment]
You can write the test [here](https://github.com/huggingface/diffusers/blob/main/tests/pipelines/stable_diffusion/test_stable_diffusion.py). Check out other tests for reference

Also this PR is not tagged with the corresponding Issue.
> Fixes # (10139)

Remove parenthesis to tag it.
Fixes #10139

**Comment by Foundsheep** [comment]
@SahilCarterr thanks. I edited the tag.

But I'm a bit confused about writing the test code. Am I supposed to add a method in the `StableDiffusionPipelineFastTests` class, and do others follow this procedure if they need to add a test case?

**Comment by SahilCarterr** [comment]
Yes, add test here `StableDiffusionPipelineFastTests` @Foundsheep

**Review by hlky — APPROVED** [review]

File: tests/pipelines/stable_diffusion/test_stable_diffusion.py [hunk 1]
```diff
@@ -840,6 +840,14 @@ def callback_on_step_end(pipe, i, t, callback_kwargs):
         # they should be the same
         assert torch.allclose(intermediate_latent, output_interrupted, atol=1e-4)
 
+    def test_pipeline_accept_tuple_type_unet_sample_size(self):
+        # the purpose of this test is to see whether the pipeline would accept a unet with the tuple-typed sample size
+        sd_repo_id = "stable-diffusion-v1-5/stable-diffusion-v1-5"
+        sample_size = [60, 80]
+        customised_unet = UNet2DConditionModel(sample_size=sample_size)
+        pipe = StableDiffusionPipeline.from_pretrained(sd_repo_id, unet=customised_unet)
+        assert pipe.unet.config.sample_size == sample_size
```
Comment (sayakpaul) [inline tests/pipelines/stable_diffusion/test_stable_diffusion.py:849]: Could we use a smaller checkpoint? 
https://huggingface.co/hf-internal-testing/tiny-sd-pipe

**Review by sayakpaul — APPROVED** [review]
Looks lit! Thank you!

**Pull request merged**